src/cc/bcc_zip.c

```diff
@@ -0,0 +1,357 @@
+  uint16_t this_disk;
+
+  // Number of the file containing the beginning of the central directory or
+  // 0xFFFF if ZIP64 archive.
+  uint16_t cd_disk;
+
+  // Number of central directory records on this disk or 0xFFFF if ZIP64
+  // archive.
+  uint16_t cd_records;
+
+  // Number of central directory records on all disks or 0xFFFF if ZIP64
+  // archive.
+  uint16_t cd_records_total;
+
+  // Size of the central directory recrod or 0xFFFFFFFF if ZIP64 archive.
+  uint32_t cd_size;
+
+  // Offset of the central directory from the beginning of the archive or
+  // 0xFFFFFFFF if ZIP64 archive.
+  uint32_t cd_offset;
+
+  // Length of comment data following end of central driectory record.
+  uint16_t comment_length;
+
+  // Up to 64k of arbitrary bytes.
+  // uint8_t comment[comment_length]
+};
+
+#define CD_FILE_HEADER_SIZE 46
+#define CD_FILE_HEADER_MAGIC 0x02014b50
+#define FLAG_ENCRYPTED (1 << 0)
+#define FLAG_HAS_DATA_DESCRIPTOR (1 << 3)
+
+struct central_directory_file_header {
+  // Magic value equal to CD_FILE_HEADER_MAGIC.
+  uint32_t magic;
+  uint16_t version;
+  // Minimum zip version needed to extract the file.
+  uint16_t min_version;
+  uint16_t flags;
+  uint16_t compression;
+  uint16_t last_modified_time;
+  uint16_t last_modified_date;
+  uint32_t crc;
+  uint32_t compressed_size;
+  uint32_t uncompressed_size;
+  uint16_t file_name_length;
+  uint16_t extra_field_length;
+  uint16_t file_comment_length;
+  // Number of the disk where the file starts or 0xFFFF if ZIP64 archive.
+  uint16_t disk;
+  uint16_t internal_attributes;
+  uint32_t external_attributes;
+  // Offset from the start of the disk containing the local file header to the
+  // start of the local file header.
+  uint32_t offset;
+};
+
+#define LOCAL_FILE_HEADER_SIZE 30
+#define LOCAL_FILE_HEADER_MAGIC 0x04034b50
+
+struct local_file_header {
+  // Magic value equal to LOCAL_FILE_HEADER_MAGIC.
+  uint32_t magic;
+  // Minimum zip version needed to extract the file.
+  uint16_t min_version;
+  uint16_t flags;
+  uint16_t compression;
+  uint16_t last_modified_time;
+  uint16_t last_modified_date;
+  uint32_t crc;
+  uint32_t compressed_size;
+  uint32_t uncompressed_size;
+  uint16_t file_name_length;
+  uint16_t extra_field_length;
+};
+
+struct bcc_zip_archive {
+  void* data;
+  uint32_t size;
+  uint32_t cd_offset;
+  uint32_t cd_records;
+};
+
+static int check_access(struct bcc_zip_archive* archive, uint32_t offset,
+                        uint32_t size) {
+  if (offset + size > archive->size || offset > offset + size) {
+    return -1;
+  }
+  return 0;
+}
+
+static int read_cd_file_header(struct bcc_zip_archive* archive,
+                               uint32_t* offset,
+                               struct central_directory_file_header* header) {
+  if (check_access(archive, *offset, CD_FILE_HEADER_SIZE)) {
+    return -1;
+  };
+
+  memcpy(header, archive->data + *offset, CD_FILE_HEADER_SIZE - 8);
+  // skip padding
```

Comment:
I can try that. I avoided packed structs in the original version of this pull request because `packed` increases complexity by an order of magnitude and might be an invitation for surprises now and in the future if this code evolves. In particular:
- I am not aware of a comprehensive list of guarantees and restrictions that reasonable compilers provide/impose on packed structs across all of the architectures we need to support. How do I ensure that misaligned accesses introduced by packed structs do not generate `SIGBUS` on arm processors ?
- it might be not enough to make all 3 structs packed. They contain 2 and 4 byte fields, but they can appear in memory in location that is 1 byte aligned. Is compiler allowed to assume that 2-byte field accesses are always aligned ? We might need to force even lower alignment.
- One of the pitfalls to avoid is capturing pointers to misaligned fields. This will be a potential danger for pull requests modifying this file in the future.
- comprehensive testing becomes much harder, we now need to test on multiple platforms.

`memcpy` is an alternative that makes the concerns above disappear. It does copy too much data though.

There is a hybrid approach. I can turn all the fields into arrays of 2 or 4 bytes and add helpers to memcpy those arrays to uint16 and uint32. As a result there should be no alignment requirements and we will not memcpy more than necessary.

Please tell me which way to go.

Reply:
> I am not aware of a comprehensive list of guarantees and restrictions that reasonable compilers provide/impose on packed structs across all of the architectures we need to support. How do I ensure that misaligned accesses introduced by packed structs do not generate SIGBUS on arm processors ?

Based on my googling, recent ARM standards (starting with [`v6`](https://patchwork.kernel.org/project/linux-arm-kernel/patch/1352205711-15787-1-git-send-email-dave.martin@linaro.org/) apparently, although I'm by no means knowledgeable wrt. ARM) won't throw on unaligned access. Am I mistaken here?

Separately, according to @yonghong-song, modern compilers should be smart enough to turn unaligned loads into single-byte loads on platforms where the former would cause an exception. 

> it might be not enough to make all 3 structs packed. They contain 2 and 4 byte fields, but they can appear in memory in location that is 1 byte aligned. Is compiler allowed to assume that 2-byte field accesses are always aligned ? We might need to force even lower alignment.

I think `__attribute__((packed, aligned(1)))` would satisfy "can appear in memory in location that is 1 byte aligned" and "no empty space between struct fields". Regardless, I think the `unaligned_uint{16,32}_t` approach you took in 2nd pass is equivalent and readable, so no need to do anything further here.

> One of the pitfalls to avoid is capturing pointers to misaligned fields. This will be a potential danger for pull requests modifying this file in the future.

I'm not too worried about this as I don't think code outside of `bcc_zip.c` will be directly interacting with these structs.

> comprehensive testing becomes much harder, we now need to test on multiple platforms.

True, but we need to test `bcc` on multiple platforms already